File: api/routes/match.js
import express from "express";
import multer from "multer";
import { getGeminiInsights } from "../services/geminiService.js";
import { parseResume } from "../services/resumeParser.js";
import { calculateMatchScore } from "../services/matchScorer.js";
import { upsertUser } from "../services/userService.js";

const router = express.Router();
const upload = multer({ dest: "uploads/" });

router.post("/match", upload.single("resume"), async (req, res) => {
  console.log("REQ comes: ", req.auth.userId);
  console.log("Session Claims:", req.auth.sessionClaims);

  if (!req.auth?.userId) {
    return res.status(401).json({ error: "Unauthorized, please sign in" });
  }

  const userId = req.auth.userId;
  const email = req.auth.sessionClaims?.primaryEmail;

  try {
    if (!email) {
      console.warn(`No email found for user ${userId}, skipping DB insert`);
    } else {
      const inserted = await upsertUser(userId, email);
      if (inserted) {
        console.log(`New user inserted: ${userId} (${email})`);
      } else {
        console.log(`User already exists: ${userId}`);
      }
    }
  } catch (err) {
    console.error("User DB error:", err.message);
    return res.status(500).json({ error: "Failed to store user" });
  }

  const { jd: jobDesc } = req.body;

  try {
    const resumeText = await parseResume(req.file.path);
    if (!resumeText || resumeText.trim().length < 20) {
      return res
        .status(400)
        .json({ error: "Resume content too short or unreadable." });
    }

    if (!jobDesc || jobDesc.trim().length < 10) {
      return res
        .status(400)
        .json({ error: "Job description is too short or missing." });
    }

    const logicScore = calculateMatchScore(resumeText, jobDesc);
    const gemini = await getGeminiInsights(resumeText, jobDesc);
    console.log("AI Missing Keywords: ", gemini.missingKeywords);

    res.json({
      logicScore: logicScore.score,
      aiScore: gemini.aiScore,
      keywordsMatched: logicScore.keywordsMatched,
      totalKeywords: logicScore.totalKeywords,
      missingKeywords: gemini.missingKeywords,
      suggestions: gemini.suggestions,
    });
  } catch (error) {
    console.error("Match error:", error.message);
    res.status(500).json({ error: "Something went wrong: " + error.message });
  }
});

export default router;

File: api_fastapi/routes/match.py
from fastapi import APIRouter, UploadFile, File, Form, HTTPException
from fastapi.responses import JSONResponse
from tempfile import NamedTemporaryFile
import shutil

from services.gemini import get_gemini_insights
from services.resumeParser import parse_resume
from services.scorer import calculate_match_score

router = APIRouter()


@router.post("/match")
async def match_resume(resume: UploadFile = File(...), jd: str = Form(...)):
    print("📥 Received /match request")

    # Save uploaded resume
    try:
        with NamedTemporaryFile(delete=False, suffix=".pdf") as tmp:
            shutil.copyfileobj(resume.file, tmp)
            temp_path = tmp.name
            print(f"📄 Resume saved to temp file: {temp_path}")
    except Exception as e:
        print("❌ Resume save error:", e)
        raise HTTPException(status_code=400, detail="Failed to save uploaded file")

    # Parse resume & match
    try:
        resume_text = parse_resume(temp_path)
        print(f"🧾 Resume parsed, length: {len(resume_text.strip())} characters")

        if not resume_text or len(resume_text.strip()) < 20:
            raise HTTPException(
                status_code=400, detail="Resume content too short or unreadable."
            )

        if not jd or len(jd.strip()) < 10:
            raise HTTPException(
                status_code=400, detail="Job description is too short or missing."
            )

        logic_score, matched_count, total_keywords = calculate_match_score(
            resume_text, jd
        )
        gemini = await get_gemini_insights(resume_text, jd)

        print("✅ Successfully processed resume and job description")

        return JSONResponse(
            content={
                "logicScore": logic_score,
                "aiScore": gemini["aiScore"],
                "keywordsMatched": matched_count,
                "totalKeywords": total_keywords,
                "missingKeywords": gemini["missingKeywords"],
                "suggestions": gemini["suggestions"],
            }
        )

    except Exception as e:
        print("❌ Match error:", str(e))
        raise HTTPException(status_code=500, detail=f"Something went wrong: {e}")
